src/mountinfo.c

```diff
@@ -0,0 +1,276 @@
+ * (9) filesystem type:  name of filesystem of the form "type[.subtype]"
+ * (10) mount source:  filesystem specific information or "none"
+ * (11) super options:  per super block options
+ **/
+static struct mountinfo_entry *parse_mountinfo_entry(const char *line)
+    __attribute__ ((nonnull(1)));
+
+/**
+ * Free a mountinfo structure and all its entries.
+ **/
+static void free_mountinfo(struct mountinfo *info)
+    __attribute__ ((nonnull(1)));
+
+/**
+ * Free a mountinfo entry.
+ **/
+static void free_mountinfo_entry(struct mountinfo_entry *entry)
+    __attribute__ ((nonnull(1)));
+
+static void cleanup_fclose(FILE ** ptr);
+static void cleanup_free(char **ptr);
+
+struct mountinfo_entry *first_mountinfo_entry(struct mountinfo *info)
+{
+	return info->first;
+}
+
+struct mountinfo_entry *next_mountinfo_entry(struct mountinfo_entry
+					     *entry)
+{
+	return entry->next;
+}
+
+int mountinfo_entry_mount_id(struct mountinfo_entry *entry)
+{
+	return entry->mount_id;
+}
+
+int mountinfo_entry_parent_id(struct mountinfo_entry *entry)
+{
+	return entry->parent_id;
+}
+
+unsigned mountinfo_entry_dev_major(struct mountinfo_entry *entry)
+{
+	return entry->dev_major;
+}
+
+unsigned mountinfo_entry_dev_minor(struct mountinfo_entry *entry)
+{
+	return entry->dev_minor;
+}
+
+const char *mountinfo_entry_root(struct mountinfo_entry *entry)
+{
+	return entry->root;
+}
+
+const char *mountinfo_entry_mount_dir(struct mountinfo_entry *entry)
+{
+	return entry->mount_dir;
+}
+
+const char *mountinfo_entry_mount_opts(struct mountinfo_entry *entry)
+{
+	return entry->mount_opts;
+}
+
+const char *mountinfo_entry_optional_fields(struct mountinfo_entry *entry)
+{
+	return entry->optional_fields;
+}
+
+const char *mountinfo_entry_fs_type(struct mountinfo_entry *entry)
+{
+	return entry->fs_type;
+}
+
+const char *mountinfo_entry_mount_source(struct mountinfo_entry *entry)
+{
+	return entry->mount_source;
+}
+
+const char *mountinfo_entry_super_opts(struct mountinfo_entry *entry)
+{
+	return entry->super_opts;
+}
+
+struct mountinfo *parse_mountinfo(const char *fname)
+{
+	if (fname == NULL) {
+		fname = "/proc/self/mountinfo";
+	}
+	FILE *f __attribute__ ((cleanup(cleanup_fclose))) = fopen(fname, "rt");
+	if (f == NULL) {
+		return NULL;
+	}
+	struct mountinfo *info = calloc(1, sizeof *info);
+	if (info == NULL) {
+		return NULL;
+	}
+	char *line __attribute__ ((cleanup(cleanup_free))) = NULL;
+	size_t line_size = 0;
+	struct mountinfo_entry *entry, *last = NULL;
+	for (;;) {
+		errno = 0;
+		if (getline(&line, &line_size, f) == -1) {
+			if (errno != 0) {
+				free_mountinfo(info);
+				return NULL;
+			}
+			break;
+		};
+		entry = parse_mountinfo_entry(line);
+		if (entry == NULL) {
+			free_mountinfo(info);
+			return NULL;
+		}
+		if (last != NULL) {
+			last->next = entry;
+		} else {
+			info->first = entry;
+		}
+		last = entry;
+	}
+	return info;
+}
+
+static struct mountinfo_entry *parse_mountinfo_entry(const char *line)
```

Comment:
As I mentioned in IRC, this function is much more complicated than I'd like it to be. However, it looks to be correct and I don't have a better suggestion at this time.